src/cml.js

```diff
@@ -0,0 +1,95 @@
+const GitlabClient = require('./GitlabClient');
+const GithubClient = require('./GithubClient');
+const { upload } = require('./utils');
+
+const get_client = (opts) => {
+  const { driver, repo, token } = opts;
+  if (!driver) throw new Error('driver not set');
+
+  if (driver === 'github') return new GithubClient({ repo, token });
+  if (driver === 'gitlab') return new GitlabClient({ repo, token });
+
+  throw new Error('driver unknown!');
+};
+
+class CML {
+  constructor(opts = {}) {
+    const env_driver = () => {
+      const { repo } = opts;
+      const { GITHUB_REPOSITORY, CI_PROJECT_URL } = process.env;
+
+      if (repo && repo.startsWith('https://github.com')) return 'github';
+      if (repo && repo.startsWith('https://gitlab.com')) return 'gitlab';
+
+      if (GITHUB_REPOSITORY) return 'github';
+      if (CI_PROJECT_URL) return 'gitlab';
+    };
+
+    const { driver = env_driver(), repo, token } = opts;
+    this.driver = driver;
+    this.repo = repo;
+    this.token = token;
+  }
+
+  env_repo() {
+    return get_client(this).env_repo();
+  }
+
+  env_token() {
+    return get_client(this).env_token();
+  }
+
+  env_is_pr() {
+    return get_client(this).env_is_pr();
+  }
+
+  env_head_sha() {
+    return get_client(this).env_head_sha();
+  }
+
+  async comment_create(opts = {}) {
+    const client = get_client(this);
+    return await client.comment_create(opts);
+  }
+
+  async check_create(opts = {}) {
+    const client = get_client(this);
+    return await client.check_create(opts);
+  }
+
+  async publish(opts = {}) {
+    const { title = '', md, gitlab_uploads } = opts;
+
+    let mime, uri;
+
+    if (gitlab_uploads) {
```

Comment:
Is there any way to move the gitlab_uploads logic out of `CML.publish()`? The only reason I ask is that if we expand to more clients with their own unique methods for uploading artifacts, I could imagine this logic getting hard to maintain.

Reply:
I think this is where it should go actually, the real logic is in the client already which is calling the gitlab api but here we define the business logic somehow you have to decide which upload is going to be used cml's native one or one provide by the client/driver

Reply:
@dmpetrov it is. The logic to publish in Gitlab is inside the Gitlab client. In CML there is only a decision. Take into account that even of you work in Gitlab you can still publish with CML storage. So its just an if

```js
if (gitlab_uploads) {
      const client = get_client({ ...this, driver: 'gitlab' });
      ({ mime, uri } = await client.publish(opts));
    } else {
      ({ mime, uri } = await upload(opts));
    }
```

If you see all the specific publish code is inside the Gitlab client

Reply:
If we were using that possibility we would be adding the  the CML upload in every client

```js
} else {
      ({ mime, uri } = await upload(opts));
    }
```